rfcs/context_uuid.md

```diff
@@ -0,0 +1,114 @@
+# RFC 87: `testdriver` Extend the mechanisms for giving browsing contexts ids.
+
+## Summary
+
+Allow contexts to be identified using either a `testdriver_id`
+property on the (window) global, or a `uuid` query parameter, either
+on the container `src` attribute (for nested browsing contexts) or the
+actual resource URL.
+
+## Details
+
+When interacting with either browsing contexts or scripting realms
+(hereafter: contexts) other than the test window, it's necessary to
+have an identifier for the context that's available both on the js
+side and can also be used by WebDriver to identify the context in
+question.
+
+Because of the limitations of the WebDriver-based testdriver setup, we
+don't have an existing UUID for each context which could be reused for
+this purpose. Instead the `testdriver-extra.js` file in wptrunner
+has a `get_window_id` function that takes a `WindowProxy` object,
+looks for a `__wptrunner_id` property, if not present sets it to a new
+UUID, and then returns the value. Then on the wptrunner side, we use
+WebDriver to search the tree of open contexts, executing script in
+each one to find the window with that property set.
+
+This setup works OK for cases where the cross-origin policy allows
+setting the property. We also support the case where the non-test
+window includes testdriver.js and can postMessage a request to the
+top-level window to initiate an action; with some work this allows
+supporting any case where it's possible to get a js handle between the
+test window and the top-level window. However we have no support for
+noopener cases or COOP headers, or other cases where it's not possible
+to pass messages between windows using js alone.
+
+To address these use cases, we propose allowing the use of a `uuid`
+query parameter on URLs, which will provide an alternate identifier
+for the window that loads the URL.
+
+In addition we propose updating the `__wptrunner_id` property name to
+`testdriver_id`, and making it a documented part of the harness
+setup. This will allow the default implementation of testdriver.js to
+refer to the property directly rather than it being part of the
+wptrunner implementation.
+
+testdriver will get a new function `get_context_id(ctx)`, with the
+following behaviour:
+* If `ctx` is a string, return `ctx`.
+* If ctx is a WindowProxy with a readable `location.href` which parses
```

Comment:
By the 2-years-in-stable policy I've suggested elsewhere, we could use `URLSearchParams`:
https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams/URLSearchParams#browser_compatibility